gui/workorder-details.lua

```diff
@@ -0,0 +1,279 @@
+    list_flags(line1, iobj.flags3)
+
+    if #line1 == 0 then
+        table.insert(line1, 'no traits')
+    end
+    return table.concat(line1, ', ')
+end
+
+function JobDetails:initListChoices()
+    if not self.job.items then
+        self.subviews.list:setChoices({})
+        return
+    end
+
+    local choices = {}
+    for i,iobj in ipairs(self.job.items) do
+        local head = 'Item '..(i+1)..' x'..iobj.quantity
+        if iobj.min_dimension > 0 then
+            head = head .. '(size '..iobj.min_dimension..')'
+        end
+
+        table.insert(choices, {
+            index = i,
+            iobj = iobj,
+            text = {
+                head, NEWLINE,
+                '  ', { text = curry(describe_item_type, iobj) }, NEWLINE,
+                '  ', { text = curry(describe_material, iobj) }, NEWLINE,
+                '  ', { text = curry(describe_item_traits, iobj) }, NEWLINE
+            }
+        })
+    end
+
+    self.subviews.list:setChoices(choices)
+end
+
+JobDetails.canChangeIType = wsj.JobDetails.canChangeIType
+JobDetails.setItemType = wsj.JobDetails.setItemType
+JobDetails.onChangeIType = wsj.JobDetails.onChangeIType
+JobDetails.canChangeMat = wsj.JobDetails.canChangeMat
+JobDetails.setMaterial = wsj.JobDetails.setMaterial
+JobDetails.findUnambiguousItem = wsj.JobDetails.findUnambiguousItem
+JobDetails.onChangeMat = wsj.JobDetails.onChangeMat
+
+function JobDetails:onInput(keys)
+    JobDetails.super.onInput(self, keys)
+end
+
+function JobDetails:canChangeTrait()
+    local idx, obj = self.subviews.list:getSelected()
+    return obj ~= nil and not is_caste_mat(obj.iobj)
+end
+
+function JobDetails:toggleFlag(obj, ffield, flag)
+    local job_item = obj.iobj
+    job_item[ffield][flag] = not job_item[ffield][flag]
+end
+
+function JobDetails:toggleToolUse(obj, tool_use)
+    local job_item = obj.iobj
+    tool_use = df.tool_uses[tool_use]
+    if job_item.has_tool_use == tool_use then
+        job_item.has_tool_use = df.tool_uses.NONE
+    else
+        job_item.has_tool_use = tool_use
+    end
+end
+
+function JobDetails:toggleMetalOre(obj, ore_ix)
+    local job_item = obj.iobj
+    if job_item.metal_ore == ore_ix then
+        job_item.metal_ore = -1
+    else
+        job_item.metal_ore = ore_ix
+    end
+end
+
+function JobDetails:toggleReactionClass(obj, reaction_class)
+    local job_item = obj.iobj
+    if job_item.reaction_class == reaction_class then
+        job_item.reaction_class = ''
+    else
+        job_item.reaction_class = reaction_class
+    end
+end
+
+function JobDetails:toggleProductMaterial(obj, product_materials)
+    local job_item = obj.iobj
+    if job_item.has_material_reaction_product == product_materials then
+        job_item.has_material_reaction_product = ''
+    else
+        job_item.has_material_reaction_product = product_materials
+    end
+end
+
+function JobDetails:unsetFlags(obj)
+    local job_item = obj.iobj
+    for flag, ffield in pairs(job_item_flags_map) do
```

Comment:
job_item_flags_map doesn't exist

Reply:
yeah... I've added a test for this, and will move the default code to the `materials.lua`

Reply:
Could you update this PR with the code that depends on DFHack/dfhack#2211? I can pull both PRs and test them together.